File: org.codefaces.ui/src/org/codefaces/ui/widgets/ajaxclient/ajax_client.js
qx.Class.define("org.codefaces.ui.widgets.ajaxclient.AjaxClient", {
  extend: qx.ui.core.Widget,
    
  construct: function() {
    this.base(arguments);
  },
  
  members: {
    
    /**
     * send an JSONP request and send back the JSONP response
     *
     * @param[String] url - the HTTP URL.
     * @param[Number] timeout - timeout in milliseconds of each request. 
     *     Use default value if set to 0.
     */
    sendJsonpRequest: function(url, timeout){
      var wid = org.eclipse.swt.WidgetManager.getInstance().findIdByWidget(this);
      
      //function for sending a reply to server
      var sendResponse = function(wid, status , content){
      var req = org.eclipse.swt.Request.getInstance();
        req.addParameter(wid + '.status', status);
        req.addParameter(wid + '.content', content);
        req.send();
      };
      
      // is the response received?
      var responseReceived = false;

      // perform the call      
	  $.jsonp({
	    url: url,
	    dataType: 'jsonp',
	    timeout: timeout,
	    callbackParameter: 'callback',
	    
	    //textStatus is always 'success'
	    success: function (data, textStatus) {
	      responseReceived = true;
	      sendResponse(wid, textStatus, JSON.stringify(data));
	    },
	    
	    //textStatus can be 'error' or 'timeout'
	    error: function(xOptions, textStatus){
         responseReceived = true;
	      sendResponse(wid, textStatus, '');
	    }
	  });
	}
  }
});

File: org.codefaces.ui/src/org/codefaces/ui/widgets/ajaxclient/AjaxClient.java
package org.codefaces.ui.widgets.ajaxclient;

import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Control;
import org.eclipse.swt.widgets.Display;

public class AjaxClient extends Control{

	private JsonpRequest jsonpRequest = null;
	private JsonpResponse jsonpResponse = null;


	public AjaxClient(Composite parent, int style) {
		super(parent, style);
	}

	/**
	 * Send the given JSONP request
	 * 
	 * @param jsonpRequest
	 *            the given JSONP request
	 * @return the response of the JSONP request
	 */
	public JsonpResponse sendJsonpRequest(JsonpRequest jsonpRequest){
		checkWidget();
		if (jsonpRequest == null) return null;
		
		this.jsonpRequest = jsonpRequest;
		this.jsonpResponse = null;
		while(this.jsonpResponse == null){
			Display display = getDisplay();
			if (!display.readAndDispatch()) {
				display.sleep();
			}
		}
		this.jsonpRequest = null;
		return this.jsonpResponse; 
	}
	
	public void setJsonpRequest(JsonpRequest request){
		jsonpRequest = request;
	}
	
	public void setJsonpResponse(JsonpResponse response){
		jsonpResponse = response;
	}
	
	public JsonpRequest getJsonpRequest() {
		return jsonpRequest;
	}

	public JsonpResponse getJsonpResponse() {
		return jsonpResponse;
	}
}
